examples/example_pp.py

```diff
@@ -0,0 +1,274 @@
+    # validate that the batch size is divisible by the microbatch_size otherwise we'll hang or error during training
+    if local_batch_size % microbatch_size != 0:
+        raise ValueError(
+            f"Batch size {local_batch_size} must be divisible by {microbatch_size=}. "
+        )
+    n_microbatches = local_batch_size // microbatch_size
+    # We expect that the number of local stages (`len(stages)`) is the same across all ranks
+    num_total_stages = pipeline_parallel_degree * len(stages)
+    if n_microbatches < num_total_stages:
+        logger.warning(
+            f"Number of microbatches ({n_microbatches}) is less than the total number "
+            f"of stages ({num_total_stages}) which may result in a bubble in the pipeline."
+        )
+
+    schedule = schedule_class(
+        stages if looped_schedule else stages[0],
+        n_microbatches=n_microbatches,
+        loss_fn=loss_fn,
+    )
+    logger.info(
+        f"Using pipeline schedule {pipeline_parallel_schedule} "
+        f"with {n_microbatches} microbatches and {num_total_stages} stages."
+    )
+    return schedule
+
+
+def run_test():
+    use_fake_pg = True
+    if not use_fake_pg:
+        # TODO(sankepurandare): Come back to this later
+        torch.distributed.init_process_group()
+        assert "WORLD_SIZE" in os.environ, "run with torchrun --nproc-per-node 4"
+        world_size = int(os.getenv("WORLD_SIZE"))
+        pp_degree = 2
+        dp_degree = 2
+        ep_degree = 2
+        assert (
+            world_size == pp_degree * dp_degree * ep_degree
+        ), "world_size must be pp * dp * ep"
+        world_mesh = torch.distributed.device_mesh.init_device_mesh(
+            "cuda",
+            (pp_degree, dp_degree, ep_degree),
+            mesh_dim_names=(
+                "pp",
+                "dp",
+                "ep",
+            ),
+        )
+        rank = int(os.getenv("RANK"))
+        local_rank = int(os.getenv("LOCAL_RANK"))
+        device = torch.device(f"cuda:{local_rank}")
+        pp_rank = world_mesh["pp"].get_local_rank()
+    else:
+        rank = int(os.getenv("RANK"))
+        pp_degree = 4
+        dp_degree = 4
+        ep_degree = 64
+        world_size = pp_degree * dp_degree * ep_degree
+
+        fake_store = FakeStore()
+        torch.distributed.init_process_group(
+            "fake", store=fake_store, rank=rank, world_size=world_size
+        )
+        # mesh = torch.distributed.device_mesh.init_device_mesh("cuda", (world_size,), mesh_dim_names=("dp",))
+        world_mesh = torch.distributed.device_mesh.init_device_mesh(
+            "cuda",
+            (pp_degree, dp_degree, ep_degree),
+            mesh_dim_names=(
+                "pp",
+                "dp",
+                "ep",
+            ),
+        )
+        device = torch.device("cuda")
+        pp_rank = rank
+
+    mesh = world_mesh["dp", "ep"]
+
+    global_batch_size = 4 * mesh.shape[1] * mesh.shape[2]
```

Comment:
huh, apparently mesh indexing doesn't work the way I thought it does (I thought this is 2D now? I guess not!)